src/include/mpir_hw_topo.h

```diff
@@ -0,0 +1,152 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef MPIR_HW_TOPO_H_INCLUDED
+#define MPIR_HW_TOPO_H_INCLUDED
+
+/*
+ * Node hardware object types
+ */
+typedef enum {
+    MPIR_NODE_HW_OBJ__MACHINE = 0,
+    MPIR_NODE_HW_OBJ__PACKAGE = 1,
+    MPIR_NODE_HW_OBJ__CORE = 2,
+    MPIR_NODE_HW_OBJ__PU = 3,
+    MPIR_NODE_HW_OBJ__L1CACHE = 4,
+    MPIR_NODE_HW_OBJ__L2CACHE = 5,
+    MPIR_NODE_HW_OBJ__L3CACHE = 6,
+    MPIR_NODE_HW_OBJ__L4CACHE = 7,
+    MPIR_NODE_HW_OBJ__L5CACHE = 8,
+    MPIR_NODE_HW_OBJ__L1ICACHE = 9,
+    MPIR_NODE_HW_OBJ__L2ICACHE = 10,
+    MPIR_NODE_HW_OBJ__L3ICACHE = 11,
+    MPIR_NODE_HW_OBJ__GROUP = 12,
+    MPIR_NODE_HW_OBJ__NUMANODE = 13,
+    MPIR_NODE_HW_OBJ__BRIDGE = 14,
+    MPIR_NODE_HW_OBJ__PCI_DEVICE = 15,
+    MPIR_NODE_HW_OBJ__OS_DEVICE = 16,
+    MPIR_NODE_HW_OBJ__MISC = 17,
+    MPIR_NODE_HW_OBJ__GPU_DEVICE = 18,
+    MPIR_NODE_HW_OBJ__NETWORK_DEVICE = 19,
+    MPIR_NODE_HW_OBJ__DDR = 20,
+    MPIR_NODE_HW_OBJ__HBM = 21,
+} MPIR_Node_hw_obj_type;
+
+/*
+ * Memory binding policies
+ */
+typedef enum {
+    MPIR_MEMBIND_POLICY__DEFAULT = 0,
+    MPIR_MEMBIND_POLICY__FIRSTTOUCH = 1,
+    MPIR_MEMBIND_POLICY__BIND = 2,
+    MPIR_MEMBIND_POLICY__INTERLEAVE = 3,
+    MPIR_MEMBIND_POLICY__NEXTTOUCH = 4,
+    MPIR_MEMBIND_POLICY__MIXED = -1
+} MPIR_Membind_policy;
+
+/*
+ * Memory binding flags
+ */
+typedef enum {
+    MPIR_MEMBIND_FLAGS__PROCESS = (1 << 0),
+    MPIR_MEMBIND_FLAGS__THREAD = (1 << 1),
+    MPIR_MEMBIND_FLAGS__STRICT = (1 << 2),
+    MPIR_MEMBIND_FLAGS__MIGRATE = (1 << 3),
+    MPIR_MEMBIND_FLAGS__NOCPUBIND = (1 << 4),
+    MPIR_MEMBIND_FLAGS__BYNODESET = (1 << 5)
+} MPIR_Membind_flags;
+
+/*
+ * Query types for network topology
+ */
+typedef enum {
+    MPIR_NETWORK_QUERY_TYPE__SWITCH_LEVEL,
+    MPIR_NETWORK_QUERY_TYPE__MIN_SIZE,
+    MPIR_NETWORK_QUERY_TYPE__MIN_MEMSIZE,
+    MPIR_NETWORK_QUERY_TYPE__TORUS_DIMENSION
+} MPIR_Network_query_type;
```

Comment:
Some of these are not network query types, e.g., `MIN_SIZE` and `MIN_MEMSIZE`.  Those are logical queries.

Reply:
Maybe I should find a better naming. These will be selecting type of queries for `netloc` so I don't have to provide three separate interfaces.

Reply:
To be clear: they are not actual queries to `netloc`.  For example, if the `MIN_SIZE` can fit inside a node, you don't need to query netloc at all.  Similarly, even if netloc is not available, these queries can be supported.  They are not directly connected to the physical topology, although their behavior would be better with physical topology information.

`TORUS_DIMENSION`, on the other hand, is directly related to network topology.

Reply:
I need to think better about how to split these to make the API as concise as possible. It seems putting this all together in the same interface will just cause confusion.

Reply:
I agree.